src/thunderbots/software/ai/hl/stp/evaluation/indirect_chip.cpp

```diff
@@ -0,0 +1,258 @@
+        {
+            if (contains(t, enemy_players[k]) == true)
+            {
+                containsEnemy = true;
+                break;
+            }
+        }
+
+        if (containsEnemy == false)
+        {
+            filtered_triangles.push_back(triangles[i]);
+        }
+    }
+
+    return filtered_triangles;
+}
+
+std::pair<Point, double> Evaluation::get_triangle_center_and_area(Triangle triangle)
+{
+    Point p1 = triangle[0];
+    Point p2 = triangle[1];
+    Point p3 = triangle[2];
+
+    double center_x = (p1.x() + p2.x() + p3.x()) / 3;
+    double center_y = (p1.y() + p2.y() + p3.y()) / 3;
+
+    double area  = abs(0.5 * ((p2.x() - p1.x()) * (p3.y() - p1.y()) -
+                             (p3.x() - p1.x()) * (p2.y() - p1.y())));
+    Point center = Point(center_x, center_y);
+
+    return std::make_pair(center, area);
+}
+
+std::vector<Triangle> Evaluation::remove_outofbounds_triangles(
+    const World& world, std::vector<Triangle> triangles)
+{
+    std::vector<Triangle> valid_triangles;
+    std::vector<Point> chip_area_corners = get_chip_target_area_corners(
+        world, Util::DynamicParameters::Indirect_Chip_Evaluation::chip_target_area_inset
+                   .value());
+    Point center;
+
+    double smallest_x = chip_area_corners[0].x();
+    double smallest_y = chip_area_corners[0].y();
+    double largest_x  = chip_area_corners[0].x();
+    double largest_y  = chip_area_corners[0].y();
+
+    for (unsigned int i = 0; i < chip_area_corners.size(); i++)
+    {
+        Point p = chip_area_corners[i];
+        if (p.x() < smallest_x)
+            smallest_x = p.x();
+        if (p.x() > largest_x)
+            largest_x = p.x();
+        if (p.y() < smallest_y)
+            smallest_y = p.y();
+        if (p.y() > largest_y)
+            largest_y = p.y();
+    }
+
+    for (unsigned int i = 0; i < triangles.size(); i++)
+    {
+        Triangle t = triangles[i];
+        center     = get_triangle_center_and_area(triangles[i]).first;
+        if (center.x() <= largest_x && center.x() >= smallest_x &&
+            center.y() <= largest_y && center.y() >= smallest_y)
+        {
+            valid_triangles.push_back(t);
+        }
+    }
+
+    return valid_triangles;
+}
+
+std::vector<Point> Evaluation::get_chip_target_area_corners(const World& world,
+                                                            double inset)
+{
+    std::vector<Point> corners;
+
+    double ballX     = world.ball().position().x();
+    double fieldX    = world.field().enemyGoal().x() - inset;
+    double negFieldY = world.field().enemyCornerNeg().y() + inset;
+    double posFieldY = world.field().enemyCornerPos().y() - inset;
+
+    Point p1 = Point(ballX, negFieldY);
+    Point p2 = Point(ballX, posFieldY);
+    Point p3 = Point(fieldX, negFieldY);
+    Point p4 = Point(fieldX, posFieldY);
+
+    corners.push_back(p1);
+    corners.push_back(p2);
+    corners.push_back(p3);
+    corners.push_back(p4);
```

Comment:
optional, but just so you know, you can write this like:
```
corners = {
   Point(ballX, negFieldY);
   Point(ballX, posFieldY);
   Point(fieldX, negFieldY);
   Point(fieldX, posFieldY);
};
return corners;
```

Reply:
Edited as per suggested. It does look cleaner than a bunch of `push_back`s! 

As a quick side question, what is the difference between `push_back` and `emplace_back`? It seems to me that using them for `std::vector` yields similar results. I tried googling and got more confused!😅

Reply:
Oh boy...... just one sec (_runs away to stackoverflow to try and understand this for the 1,000,000 time_). **OK**, I _think_ I got this. Basically, `std::push_back` does an extra `move` operation, so it's _slightly_ slower. It also doesn't handle some advanced c++ features properly, but people screwed up implementing it the first time, so it's probably here to stay.

TL;DR: use `emplace_back`, never `push_back`, though you'll probably never notice the differences

Reply:
Noted!!! Haha, tyvm for the explanation! :smile: